### PR TITLE
Add tilize / untilize ops to the ttkernel dialect

### PR DESCRIPTION
**Pull request by nsmithtt**

### CODE REVIEW

**Comment by nsmithtt** [comment]
ping

File: lib/Dialect/TTKernel/IR/TTKernelOps.cpp [hunk 1]
```diff
@@ -91,6 +91,78 @@ ::mlir::LogicalResult CBWaitFrontOp::verify() {
   return success();
 }
 
+::mlir::LogicalResult TilizeInitOp::verify() {
+  if (not insideDispatchOpRegion(getOperation())) {
```
Comment (rpavlovicTT) [inline lib/Dialect/TTKernel/IR/TTKernelOps.cpp:95]: Can we enforce a rule to use `!` for negation and forbid `not` please?

File: lib/Dialect/TTKernel/IR/TTKernelOps.cpp [hunk 2]
```diff
@@ -91,6 +91,78 @@ ::mlir::LogicalResult CBWaitFrontOp::verify() {
...
+    return emitOpError("TilizeInitOp must be inside of a DispatchOp region");
+  }
+  if (getCbIn().getType().getPort() == getCbOut().getType().getPort()) {
+    return emitOpError("cbIn and cbOut must be different");
```
Comment (rpavlovicTT) [inline lib/Dialect/TTKernel/IR/TTKernelOps.cpp:99]: I'd be more clear when emitting errors, for example here I'd say:
`Input circular buffer port and output circular buffer port must be different`

Same for the remainder of errors in the file.

File: lib/Dialect/TTKernel/IR/TTKernelOps.cpp [hunk 3]
```diff
@@ -91,6 +91,78 @@ ::mlir::LogicalResult CBWaitFrontOp::verify() {
...
+  if (not mlir::isa<tt::TileType>(
+          getCbOut().getType().getMemref().getElementType())) {
+    return emitOpError("cbOut must have tile element type");
+  }
+  return success();
+}
+
+::mlir::LogicalResult UntilizeBlockOp::verify() {
+  if (not insideDispatchOpRegion(getOperation())) {
+    return emitOpError("TilizeBlockOp must be inside of a DispatchOp region");
```
Comment (rpavlovicTT) [inline lib/Dialect/TTKernel/IR/TTKernelOps.cpp:150]: UntilizeBlockOp*

File: lib/Dialect/TTKernel/IR/TTKernelOps.cpp [hunk 4]
```diff
@@ -91,6 +91,78 @@ ::mlir::LogicalResult CBWaitFrontOp::verify() {
...
+  if (not mlir::isa<tt::TileType>(
+          getCbOut().getType().getMemref().getElementType())) {
+    return emitOpError("cbOut must have tile element type");
+  }
+  return success();
+}
+
+::mlir::LogicalResult UntilizeInitOp::verify() {
+  if (not insideDispatchOpRegion(getOperation())) {
+    return emitOpError("TilizeInitOp must be inside of a DispatchOp region");
```
Comment (rpavlovicTT) [inline lib/Dialect/TTKernel/IR/TTKernelOps.cpp:114]: UntilizeInitOp*

File: lib/Dialect/TTKernel/IR/TTKernelOps.cpp [hunk 5]
```diff
@@ -91,6 +91,78 @@ ::mlir::LogicalResult CBWaitFrontOp::verify() {
...
+  }
+  if (not mlir::isa<tt::TileType>(
+          getCbIn().getType().getMemref().getElementType())) {
+    return emitOpError("cbIn must have tile element type");
+  }
+  if (mlir::isa<tt::TileType>(
+          getCbOut().getType().getMemref().getElementType())) {
+    return emitOpError("cbOut must have scalar element type");
+  }
+  return success();
```
Comment (rpavlovicTT) [inline lib/Dialect/TTKernel/IR/TTKernelOps.cpp:163]: This whole block repeats for UntilizeBlockOp and UntilizeInitOp. Let's extract it to a function. Same can be done for TilizeInitOp & TilizeBlockop.

**Review by rpavlovicTT — CHANGES_REQUESTED** [review]

**Review by rpavlovicTT — APPROVED** [review]

**Pull request merged**